Support: std::randomize() fails to compile for queue and associative array variables

`std::randomize()` causes C++ compilation errors when called on queue (`$`), dynamic array (`[]`), or associative array variables. The `VlStdRandomizer::basicStdRandomization()` template only has specializations for scalar, wide, and `VlUnpacked` types — missing overloads for `VlQueue` and `VlAssocArray`.


```systemverilog
module t;
  bit [7:0] q [$];
  initial begin
    q = {8'd0, 8'd0, 8'd0};
    if (std::randomize(q) != 1) $stop;
    $finish;
  end
endmodule

```

### Error

verilated_random.h:622: error: no match for 'operator=' (operand types are 'VlQueue<unsigned char>' and 'unsigned int')
The scalar specialization is selected instead of iterating queue elements.

### Affected types
VlQueue<T, N> — queues and dynamic arrays
VlAssocArray<K, V> — associative arrays
Fixed-size unpacked arrays (VlUnpacked) work correctly (specialization exists at line 639).

### IEEE reference
IEEE 1800-2017 §18.5 — std::randomize() should support all variable types including arrays and queues.

### Source
From [PlanV_Verilator_Feature_Tests#33](https://github.com/planvtech/PlanV_Verilator_Feature_Tests/issues/33) and [#66](https://github.com/planvtech/PlanV_Verilator_Feature_Tests/issues/66).

A PR is coming. ;)

--- a/include/verilated_random.h
+++ b/include/verilated_random.h
@@ -641,6 +641,23 @@ public:
         for (size_t i = 0; i < N_Depth; ++i) { basicStdRandomization(value.operator[](i), width); }
         return true;
     }
+
+    // Queue/dynamic array randomization
+    template <typename T_Value, size_t N_MaxSize>
+    bool basicStdRandomization(VlQueue<T_Value, N_MaxSize>& value, size_t width) {
+        for (int i = 0; i < value.size(); ++i) { basicStdRandomization(value.atWrite(i), width); }
+        return true;
+    }
+
+    // Associative array randomization
+    template <typename T_Key, typename T_Value>
+    bool basicStdRandomization(VlAssocArray<T_Key, T_Value>& value, size_t width) {
+        T_Key key;
+        for (int exists = value.first(key); exists; exists = value.next(key)) {
+            basicStdRandomization(value.at(key), width);
+        }
+        return true;
+    }
     bool next() { return VlRandomizer::next(m_rng); }
 };
 

--- a/test_regress/t/t_std_randomize_assoc.py
+++ b/test_regress/t/t_std_randomize_assoc.py
@@ -1,0 +1,21 @@
+#!/usr/bin/env python3
+# DESCRIPTION: Verilator: Verilog Test driver/expect definition
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of either the GNU Lesser General Public License Version 3
+# or the Perl Artistic License Version 2.0.
+# SPDX-FileCopyrightText: 2026 Wilson Snyder
+# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+
+import vltest_bootstrap
+
+test.scenarios('simulator')
+
+if not test.have_solver:
+    test.skip("No constraint solver installed")
+
+test.compile()
+
+test.execute()
+
+test.passes()

--- a/test_regress/t/t_std_randomize_assoc.v
+++ b/test_regress/t/t_std_randomize_assoc.v
@@ -1,0 +1,48 @@
+// DESCRIPTION: Verilator: Verilog Test module
+//
+// This file ONLY is placed under the Creative Commons Public Domain.
+// SPDX-FileCopyrightText: 2026 PlanV GmbH
+// SPDX-License-Identifier: CC0-1.0
+
+// Test std::randomize() with associative array variables
+
+`define stop $stop;
+`define checkd(gotv,expv) do if ((gotv) !== (expv)) begin $write("%%Error: %s:%0d:  got=%0d exp=%0d\n", `__FILE__,`__LINE__, (gotv), (expv)); `stop; end while(0);
+
+module t_std_randomize_assoc;
+
+  // Associative array variables
+  int assoc_int [int];
+  bit [7:0] assoc_byte [string];
+
+  initial begin
+    // Test 1: std::randomize with int-keyed associative array (no constraints)
+    assoc_int[0] = 0;
+    assoc_int[1] = 0;
+    assoc_int[2] = 0;
+    `checkd(std::randomize(assoc_int), 1);
+
+    // Test 2: std::randomize with string-keyed associative array
+    assoc_byte["a"] = 0;
+    assoc_byte["b"] = 0;
+    assoc_byte["c"] = 0;
+    `checkd(std::randomize(assoc_byte), 1);
+
+    // Test 3: Multiple randomizations produce different values
+    begin
+      automatic int non_zero = 0;
+      repeat (5) begin
+        assoc_int[0] = 0;
+        assoc_int[1] = 0;
+        `checkd(std::randomize(assoc_int), 1);
+        if (assoc_int[0] != 0) non_zero++;
+        if (assoc_int[1] != 0) non_zero++;
+      end
+      // With 10 random int values, expect most non-zero
+      if (non_zero < 7) `stop;
+    end
+
+    $write("*-* All Finished *-*\n");
+    $finish;
+  end
+endmodule

--- a/test_regress/t/t_std_randomize_queue.py
+++ b/test_regress/t/t_std_randomize_queue.py
@@ -1,0 +1,21 @@
+#!/usr/bin/env python3
+# DESCRIPTION: Verilator: Verilog Test driver/expect definition
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of either the GNU Lesser General Public License Version 3
+# or the Perl Artistic License Version 2.0.
+# SPDX-FileCopyrightText: 2026 Wilson Snyder
+# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+
+import vltest_bootstrap
+
+test.scenarios('simulator')
+
+if not test.have_solver:
+    test.skip("No constraint solver installed")
+
+test.compile()
+
+test.execute()
+
+test.passes()

--- a/test_regress/t/t_std_randomize_queue.v
+++ b/test_regress/t/t_std_randomize_queue.v
@@ -1,0 +1,70 @@
+// DESCRIPTION: Verilator: Verilog Test module
+//
+// This file ONLY is placed under the Creative Commons Public Domain.
+// SPDX-FileCopyrightText: 2026 PlanV GmbH
+// SPDX-License-Identifier: CC0-1.0
+
+// Test std::randomize() with queue and dynamic array variables
+
+`define stop $stop;
+`define checkd(gotv,expv) do if ((gotv) !== (expv)) begin $write("%%Error: %s:%0d:  got=%0d exp=%0d\n", `__FILE__,`__LINE__, (gotv), (expv)); `stop; end while(0);
+
+module t_std_randomize_queue;
+
+  // Queue variables
+  bit [7:0] q8 [$];
+  bit [31:0] q32 [$];
+
+  // Dynamic array variables
+  bit [7:0] dyn8 [];
+
+  int i;
+
+  initial begin
+    // Test 1: std::randomize with queue (no constraints)
+    q8 = {8'd0, 8'd0, 8'd0};
+    `checkd(std::randomize(q8), 1);
+
+    // Test 2: std::randomize with queue and constraints
+    q8 = {8'd0, 8'd0, 8'd0, 8'd0};
+    `checkd(std::randomize(q8) with {
+      foreach (q8[j]) q8[j] > 8'd10 && q8[j] < 8'd100;
+    }, 1);
+    foreach (q8[j]) begin
+      if (q8[j] <= 8'd10 || q8[j] >= 8'd100) `stop;
+    end
+
+    // Test 3: std::randomize with 32-bit queue
+    q32 = {32'd0, 32'd0, 32'd0};
+    `checkd(std::randomize(q32) with {
+      foreach (q32[k]) q32[k] < 32'd1000;
+    }, 1);
+    foreach (q32[k]) begin
+      if (q32[k] >= 32'd1000) `stop;
+    end
+
+    // Test 4: std::randomize with dynamic array
+    dyn8 = new[3];
+    `checkd(std::randomize(dyn8) with {
+      foreach (dyn8[m]) dyn8[m] inside {[1:50]};
+    }, 1);
+    foreach (dyn8[m]) begin
+      if (dyn8[m] < 1 || dyn8[m] > 50) `stop;
+    end
+
+    // Test 5: Multiple randomizations produce different values
+    q8 = {8'd0, 8'd0, 8'd0};
+    begin
+      automatic int non_zero = 0;
+      repeat (5) begin
+        `checkd(std::randomize(q8), 1);
+        foreach (q8[n]) if (q8[n] != 0) non_zero++;
+      end
+      // With 15 random 8-bit values, expect most non-zero
+      if (non_zero < 10) `stop;
+    end
+
+    $write("*-* All Finished *-*\n");
+    $finish;
+  end
+endmodule

